transport/src/main/java/io/netty/channel/StacklessClosedChannelException.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel;
+
+import io.netty.util.internal.ThrowableUtil;
+
+import java.nio.channels.ClosedChannelException;
+
+/**
+ * Cheap {@link ClosedChannelException} that does not fill in the stacktrace.
+ */
+final class StacklessClosedChannelException extends ClosedChannelException {
+
+    private static final long serialVersionUID = -2214806025529435136L;
+
+    private StacklessClosedChannelException() { }
+
+    @Override
+    public Throwable fillInStackTrace() {
+        return this;
+    }
+
+    /**
+     * Creates a new {@link StacklessClosedChannelException} which has the origin of the given {@link Class} and method.
+     */
+    static StacklessClosedChannelException newInstance(Class<?> clazz, String method) {
+        return ThrowableUtil.unknownStackTrace(new StacklessClosedChannelException(), clazz, method);
+    }
```

Comment:
- Could optimize even further by pre-constructing some instances and reusing them. To make it truly immutable, you have to set `false` to `enableSuppression` and `writableStackTrace` by calling the protected super constructor at `Throwable`.
- Would be nice if a static factory method is moved before the member fields for consistency.

Reply:
See [here](https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/common/ClosedSessionException.java) for a complete example, which uses a pre-instantiated singleton for a very common situation, while creating a new instance for others.

Reply:
That being said, how about just merging this into `ClosedChannelException`? Adding a static factory method there wouldn't hurt backward compatibility.

Reply:
@trustin using a static instance is very problematic and may cause OOME's because users may call `addSuppressed(...)` on it. This can then lead to an ever-growing instance that at the end cause an OOME. So I highly recommend fix this in Armeria  ;)

Reply:
Also note that `enableSuppression` is not exposed in `CloseChannelException`

Reply:
> users may call addSuppressed(...) on it. 

Setting `false` to `enableSuppression` prevents it, so nothing to fix in Armeria fortunately 😉 

> Also note that enableSuppression is not exposed in CloseChannelException

Doh! Missed that 😆

Reply:
Just remembered `ClosedChannelException` is a JDK class. Lack of sleep really hurts 😅